File: android/app/src/main/java/com/skdaddle/spellit/ui/GameScaffold.kt
package com.skdaddle.spellit.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.outlined.StarOutline
import androidx.compose.material3.Icon
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.key
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.isSpecified
import androidx.compose.ui.unit.sp
import kotlin.math.ceil
import com.skdaddle.spellit.engine.RoundEngine
import com.skdaddle.spellit.model.WordEntry

// Score bar

@Composable
fun ScoreBar(
    unit: String = "Word",
    index: Int,
    total: Int,
    score: Int,
    streak: Int,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                "$unit ${minOf(index + 1, total)} of $total",
                style = headingStyle(13, FontWeight.Medium),
                color = Palette.MutedInk,
            )
            Spacer(Modifier.weight(1f))
            if (streak >= 2) {
                Icon(
                    Icons.Filled.LocalFireDepartment,
                    contentDescription = null,
                    tint = Palette.Coral,
                    modifier = Modifier.size(16.dp),
                )
                Text(
                    " $streak in a row!  ",
                    style = headingStyle(13, FontWeight.Medium),
                    color = Palette.Coral,
                )
            }
            Icon(
                Icons.Filled.Star,
                contentDescription = null,
                tint = Palette.Sun,
                modifier = Modifier.size(16.dp),
            )
            Text(
                " $score",
                style = headingStyle(13, FontWeight.Medium),
                color = Palette.Ink,
            )
        }
        LinearProgressIndicator(
            progress = { index.toFloat() / maxOf(total, 1) },
            color = Palette.Ink,
            trackColor = Palette.SecondaryBg,
            modifier = Modifier
                .fillMaxWidth()
                .clip(CircleShape),
        )
    }
}

// Round summary

@OptIn(ExperimentalLayoutApi::class)
@Composable
fun RoundSummary(
    score: Int,
    total: Int,
    bestStreak: Int,
    missed: List<WordEntry>,
    summaryText: String? = null,
    onRestart: () -> Unit,
) {
    val stars = RoundEngine.stars(score, total)
    val headline = when (stars) {
        3 -> "Wow! Spelling superstar!"
        2 -> "Nice spelling! One more round?"
        1 -> "Good effort — try for more stars!"
        else -> "Keep practicing — you'll get there!"
    }
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 8.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
            for (i in 0 until 3) {
                val shape = RoundedCornerShape(14.dp)
                Box(
                    Modifier
                        .size(56.dp)
                        .rotate(if (i == 0) -6f else if (i == 2) 6f else 0f),
                ) {
                    Box(
                        Modifier
                            .matchParentSize()
                            .offset(y = 4.dp)
                            .background(Palette.Ink, shape),
                    )
                    Box(
                        Modifier
                            .matchParentSize()
                            .background(if (i < stars) Palette.SunSoft else Palette.SecondaryBg, shape)
                            .border(3.dp, Palette.Ink, shape),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(
                            if (i < stars) Icons.Filled.Star else Icons.Outlined.StarOutline,
                            contentDescription = null,
                            tint = if (i < stars) Palette.Sun else Palette.MutedInk,
                            modifier = Modifier.size(30.dp),
                        )
                    }
                }
            }
        }

        Text(
            headline,
            style = headingStyle(22),
            color = Palette.Ink,
            textAlign = TextAlign.Center,
        )

        val defaultSummary = buildString {
            append("You spelled $score of $total words right")
            if (bestStreak >= 3) append(" — best streak: $bestStreak in a row")
            append(".")
        }
        Text(
            summaryText ?: defaultSummary,
            fontSize = 15.sp,
            color = Palette.MutedInk,
            textAlign = TextAlign.Center,
        )

        if (missed.isNotEmpty()) {
            Column(
                Modifier
                    .fillMaxWidth()
                    .background(Palette.SecondaryBg, RoundedCornerShape(14.dp))
                    .padding(14.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Text(
                    "Words to practice",
                    style = headingStyle(13, FontWeight.Medium),
                    color = Palette.MutedInk,
                )
                FlowRow(
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    for (entry in missed) {
                        Text(
                            entry.word,
                            style = headingStyle(14, FontWeight.Medium),
                            color = Palette.Ink,
                            modifier = Modifier
                                .background(Color.White, RoundedCornerShape(10.dp))
                                .border(2.dp, Palette.Ink, RoundedCornerShape(10.dp))
                                .padding(horizontal = 10.dp, vertical = 5.dp),
                        )
                    }
                }
            }
        }

        ChunkyButton(text = "Play again", onClick = onRestart)
    }
}

// Feedback panel (correct / reveal)

@Composable
fun FeedbackPanel(
    correct: Boolean,
    word: String,
    isLast: Boolean,
    onNext: () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .background(
                if (correct) Palette.LeafSoft else Palette.CoralSoft,
                RoundedCornerShape(16.dp),
            )
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        if (correct) {
            Text("Nailed it! ⭐️", style = headingStyle(19), color = Palette.Ink)
        } else {
            Text("Almost! It's spelled:", style = headingStyle(17), color = Palette.Ink)
            WordTiles(word = word)
        }
        // A same-frame double-tap must not advance twice (skipping a word);
        // the panel leaves composition on advance, so this resets per word.
        val advanced = androidx.compose.runtime.remember {
            androidx.compose.runtime.mutableStateOf(false)
        }
        ChunkyButton(
            text = if (isLast) "See my score" else "Next word",
            onClick = {
                if (!advanced.value) {
                    advanced.value = true
                    onNext()
                }
            },
        )
    }
}

// Fitting a word onto one line

/** How a run of tiles should sit inside a measured width. */
data class TileFit(val size: TileSize, val spacing: Dp, val perRow: Int)

/**
 * Fixed tile sizes wrap a word wherever the row happens to run out of room,
 * which leaves orphan letters on the next line and is hard to read. A fit
 * shrinks the tiles so the whole word stays on one line, and only splits — into
 * even rows — once that would push the letters below a legible size, so
 * "pronunciation" reads as 7 + 6 rather than 9 + 4.
 */
object TileLadder {
    /** Full-size tile; matches [TileSize.MD]. */
    private val MaxSide = 48.dp
    /** Smallest tile a young reader can still read comfortably. */
    private val FloorSide = 28.dp
    /** Only reached when even three rows will not fit. */
    private val HardMinSide = 18.dp
    /** Gaps tighten alongside the tiles so narrow screens buy back some room. */
    private val RoomySpacing = 6.dp
    private val TightSpacing = 4.dp
    /** Past this, splitting the word does more harm than shrinking it. */
    private const val MaxRows = 3

    /**
     * Lays [count] tiles into [width], reserving [extra] for anything else
     * sharing the line (the "+" and "=" in Ending Machine).
     */
    fun fit(count: Int, width: Dp, extra: Dp = 0.dp): TileFit {
        if (count <= 0) return TileFit(TileSize.MD, RoomySpacing, 1)
        // Before the row is measured, fall back to the static heuristic.
        if (!width.isSpecified || !width.value.isFinite() || width <= 0.dp) {
            return TileFit(TileSize.forCount(count), RoomySpacing, count)
        }
        for (rows in 1..MaxRows) {
            val perRow = ceil(count.toDouble() / rows).toInt()
            val (side, spacing) = sideFor(perRow, width, extra)
            if (side >= FloorSide || rows == MaxRows) {
                return TileFit(TileSize.of(maxOf(HardMinSide, side)), spacing, perRow)
            }
        }
        return TileFit(TileSize.forCount(count), RoomySpacing, count)
    }

    private fun sideFor(perRow: Int, width: Dp, extra: Dp): Pair<Dp, Dp> {
        val roomy = (width - extra - RoomySpacing * (perRow - 1)) / perRow
        if (roomy >= 36.dp) return minOf(MaxSide, roomy) to RoomySpacing
        val tight = (width - extra - TightSpacing * (perRow - 1)) / perRow
        return minOf(MaxSide, tight) to TightSpacing
    }
}

/**
 * A run of tiles sized to fit the width it is given, split into even rows only
 * when the tiles would otherwise be too small to read.
 */
@Composable
fun TileRow(
    count: Int,
    modifier: Modifier = Modifier,
    tile: @Composable (index: Int, size: TileSize) -> Unit,
) {
    BoxWithConstraints(modifier.fillMaxWidth()) {
        val fit = TileLadder.fit(count, maxWidth)
        Column(
            modifier = Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(fit.spacing),
        ) {
            for (start in 0 until count step fit.perRow) {
                key(start) {
                    Row(horizontalArrangement = Arrangement.spacedBy(fit.spacing)) {
                        for (index in start until minOf(start + fit.perRow, count)) {
                            key(index) { tile(index, fit.size) }
                        }
                    }
                }
            }
        }
    }
}

/** A word rendered as a row of tiles that fits the space it is given. */
@Composable
fun WordTiles(
    word: String,
    modifier: Modifier = Modifier,
    fill: Color = Color.White,
) {
    TileRow(count = word.length, modifier = modifier) { index, size ->
        Tile(letter = word[index].toString(), size = size, fill = fill)
    }
}

// Game scaffold

@Composable
fun GameScaffold(
    game: Game,
    engine: RoundEngine,
    unit: String = "Word",
    summaryText: String? = null,
    onRestart: () -> Unit,
    picker: @Composable () -> Unit,
    content: @Composable () -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Palette.Paper)
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(18.dp),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(14.dp),
        ) {
            val shape = RoundedCornerShape(14.dp)
            Box(Modifier.size(54.dp)) {
                Box(
                    Modifier
                        .matchParentSize()
                        .offset(y = 4.dp)
                        .background(Palette.Ink, shape),
                )
                Box(
                    Modifier
                        .matchParentSize()
                        .background(game.accentSoft, shape)
                        .border(3.dp, Palette.Ink, shape),
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(
                        game.icon,
                        contentDescription = null,
                        tint = Palette.Ink,
                        modifier = Modifier.size(28.dp),
                    )
                }
            }
            Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text(game.title, style = headingStyle(26), color = Palette.Ink)
                Text(game.instructions, fontSize = 13.sp, color = Palette.MutedInk)
            }
        }

        picker()

        if (engine.phase == RoundEngine.Phase.PLAYING && engine.words.isNotEmpty()) {
            ScoreBar(
                unit = unit,
                index = engine.index,
                total = engine.words.size,
                score = engine.score,
                streak = engine.streak,
            )
        }

        Box(
            Modifier
                .fillMaxWidth()
                .shadow(4.dp, RoundedCornerShape(18.dp))
                .background(Color.White, RoundedCornerShape(18.dp)),
        ) {
            Box(
                Modifier
                    .fillMaxWidth()
                    .height(7.dp)
                    .background(
                        game.accent,
                        RoundedCornerShape(topStart = 18.dp, topEnd = 18.dp),
                    )
                    .align(Alignment.TopCenter),
            )
            Box(Modifier.padding(18.dp)) {
                if (engine.phase == RoundEngine.Phase.DONE) {
                    RoundSummary(
                        score = engine.score,
                        total = engine.words.size,
                        bestStreak = engine.bestStreak,
                        missed = engine.missedWords,
                        summaryText = summaryText,
                        onRestart = onRestart,
                    )
                } else {
                    content()
                }
            }
        }
    }
}

File: ios/SpellIt/Views/GameScaffold.swift
import SwiftUI

// MARK: - Grade picker

struct GradePicker: View {
    @Binding var grade: GradeBand

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Pick your level")
                .font(.heading(13, weight: .medium))
                .foregroundStyle(Color.mutedInk)
            HStack(spacing: 8) {
                ForEach(GradeBand.allCases) { band in
                    Button {
                        grade = band
                    } label: {
                        Text(band.label)
                            .font(.heading(13, weight: .medium))
                            .padding(.horizontal, 10)
                            .padding(.vertical, 7)
                            .background(
                                Capsule().fill(grade == band ? Color.ink : Color.white)
                            )
                            .overlay(Capsule().strokeBorder(Color.ink, lineWidth: 2.5))
                            .foregroundStyle(grade == band ? Color.white : Color.ink)
                    }
                    .accessibilityAddTraits(grade == band ? .isSelected : [])
                }
            }
            Text(grade.blurb)
                .font(.system(size: 13))
                .foregroundStyle(Color.mutedInk)
        }
    }
}

// MARK: - Score bar

struct ScoreBar: View {
    var unit = "Word"
    var index: Int
    var total: Int
    var score: Int
    var streak: Int

    var body: some View {
        VStack(spacing: 6) {
            HStack {
                Text("\(unit) \(min(index + 1, total)) of \(total)")
                Spacer()
                if streak >= 2 {
                    Label("\(streak) in a row!", systemImage: "flame.fill")
                        .foregroundStyle(Color.coral)
                }
                Label("\(score)", systemImage: "star.fill")
                    .foregroundStyle(Color.ink)
                    .symbolRenderingMode(.multicolor)
            }
            .font(.heading(13, weight: .medium))
            .foregroundStyle(Color.mutedInk)
            ProgressView(value: Double(index), total: Double(max(total, 1)))
                .tint(Color.ink)
        }
    }
}

// MARK: - Round summary

struct RoundSummaryView: View {
    var score: Int
    var total: Int
    var bestStreak: Int
    var missed: [WordEntry]
    var summaryText: String?
    var onRestart: () -> Void

    private var stars: Int { RoundEngine.stars(score: score, total: total) }

    private var headline: String {
        switch stars {
        case 3: "Wow! Spelling superstar!"
        case 2: "Nice spelling! One more round?"
        case 1: "Good effort — try for more stars!"
        default: "Keep practicing — you'll get there!"
        }
    }

    var body: some View {
        VStack(spacing: 16) {
            HStack(spacing: 10) {
                ForEach(0..<3, id: \.self) { i in
                    ZStack {
                        RoundedRectangle(cornerRadius: 14)
                            .fill(Color.ink)
                            .offset(y: 4)
                        RoundedRectangle(cornerRadius: 14)
                            .fill(i < stars ? Color.sunSoft : Color.secondaryBg)
                            .overlay(RoundedRectangle(cornerRadius: 14).strokeBorder(Color.ink, lineWidth: 3))
                        Image(systemName: i < stars ? "star.fill" : "star")
                            .font(.system(size: 26))
                            .foregroundStyle(i < stars ? Color.sun : Color.mutedInk)
                    }
                    .frame(width: 56, height: 56)
                    .rotationEffect(.degrees(i == 0 ? -6 : i == 2 ? 6 : 0))
                }
            }
            .accessibilityElement(children: .ignore)
            .accessibilityLabel("\(stars) of 3 stars")

            Text(headline)
                .font(.heading(22, weight: .semibold))
                .foregroundStyle(Color.ink)
                .multilineTextAlignment(.center)

            Text(summaryText ?? defaultSummary)
                .font(.system(size: 15))
                .foregroundStyle(Color.mutedInk)
                .multilineTextAlignment(.center)

            if !missed.isEmpty {
                VStack(alignment: .leading, spacing: 8) {
                    Text("Words to practice")
                        .font(.heading(13, weight: .medium))
                        .foregroundStyle(Color.mutedInk)
                    FlowLayout(spacing: 8) {
                        ForEach(missed) { entry in
                            Text(entry.word)
                                .font(.heading(14, weight: .medium))
                                .padding(.horizontal, 10)
                                .padding(.vertical, 5)
                                .background(RoundedRectangle(cornerRadius: 10).fill(Color.white))
                                .overlay(RoundedRectangle(cornerRadius: 10).strokeBorder(Color.ink, lineWidth: 2))
                        }
                    }
                }
                .padding(14)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(RoundedRectangle(cornerRadius: 14).fill(Color.secondaryBg))
            }

            Button("Play again", action: onRestart)
                .buttonStyle(ChunkyButtonStyle())
        }
        .padding(.vertical, 8)
    }

    private var defaultSummary: String {
        var text = "You spelled \(score) of \(total) words right"
        if bestStreak >= 3 { text += " — best streak: \(bestStreak) in a row" }
        return text + "."
    }
}

// MARK: - Feedback panel (correct / reveal)

struct FeedbackPanel: View {
    var correct: Bool
    var word: String
    var isLast: Bool
    var onNext: () -> Void

    var body: some View {
        VStack(spacing: 12) {
            if correct {
                Text("Nailed it! ⭐️")
                    .font(.heading(19, weight: .semibold))
                    .foregroundStyle(Color.ink)
            } else {
                Text("Almost! It's spelled:")
                    .font(.heading(17, weight: .semibold))
                    .foregroundStyle(Color.ink)
                WordTilesView(word: word)
            }
            Button(isLast ? "See my score" : "Next word", action: onNext)
                .buttonStyle(ChunkyButtonStyle())
        }
        .padding(16)
        .frame(maxWidth: .infinity)
        .background(
            RoundedRectangle(cornerRadius: 16)
                .fill(correct ? Color.leafSoft : Color.coralSoft)
        )
        .accessibilityElement(children: .combine)
    }
}

// MARK: - Fitting a word onto one line

/// Fixed tile sizes wrap a word wherever the row happens to run out of room,
/// which leaves orphan letters on the next line and is hard to read. These
/// candidates are offered to `ViewThatFits` largest first: shrink the tiles to
/// keep the whole word on one line, and only once they would be too small for a
/// young reader split into even rows, so "pronunciation" reads as 7 + 6 rather
/// than 9 + 4.
enum TileLadder {
    /// Full-size tile, matching `.md`, down to the smallest a child can still
    /// read comfortably.
    static let sides: [CGFloat] = [48, 40, 34, 28]
    /// Last resort, when even three rows of the smallest readable tile overflow.
    static let hardMinSide: CGFloat = 22

    /// Gaps tighten alongside the tiles so narrow screens buy back some room.
    static func spacing(for side: CGFloat) -> CGFloat { side < 36 ? 4 : 6 }

    /// Even rows: 13 tiles over 2 rows is 7 + 6, never 7 + 6 reordered.
    static func rows(count: Int, over rowCount: Int) -> [Range<Int>] {
        guard count > 0, rowCount > 0 else { return [] }
        let perRow = Int((Double(count) / Double(rowCount)).rounded(.up))
        return stride(from: 0, to: count, by: perRow).map {
            $0..<min($0 + perRow, count)
        }
    }
}

/// A run of tiles sized to fit the width it is given, split into even rows only
/// when the tiles would otherwise be too small to read.
struct TileRow<Tile: View>: View {
    var count: Int
    @ViewBuilder var tile: (Int, TileSize) -> Tile

    var body: some View {
        ViewThatFits(in: .horizontal) {
            candidate(rows: 1, side: TileLadder.sides[0])
            candidate(rows: 1, side: TileLadder.sides[1])
            candidate(rows: 1, side: TileLadder.sides[2])
            candidate(rows: 1, side: TileLadder.sides[3])
            candidate(rows: 2, side: TileLadder.sides[0])
            candidate(rows: 2, side: TileLadder.sides[1])
            candidate(rows: 2, side: TileLadder.sides[2])
            candidate(rows: 2, side: TileLadder.sides[3])
            candidate(rows: 3, side: TileLadder.sides[2])
            candidate(rows: 3, side: TileLadder.hardMinSide)
        }
    }

    @ViewBuilder
    private func candidate(rows rowCount: Int, side: CGFloat) -> some View {
        let spacing = TileLadder.spacing(for: side)
        let size = TileSize(side: side)
        VStack(spacing: spacing) {
            ForEach(TileLadder.rows(count: count, over: rowCount), id: \.lowerBound) { range in
                HStack(spacing: spacing) {
                    ForEach(range, id: \.self) { index in
                        tile(index, size)
                    }
                }
            }
        }
    }
}

/// A word rendered as a row of tiles that fits the space it is given.
struct WordTilesView: View {
    var word: String
    var fill: Color = .white

    var body: some View {
        let letters = Array(word)
        TileRow(count: letters.count) { index, size in
            TileView(letter: String(letters[index]), size: size, fill: fill)
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(word.map(String.init).joined(separator: " "))
    }
}

// MARK: - Game scaffold

struct GameScaffold<Content: View, Picker: View>: View {
    var game: Game
    var engine: RoundEngine
    var unit = "Word"
    var summaryText: String? = nil
    var onRestart: () -> Void
    @ViewBuilder var picker: () -> Picker
    @ViewBuilder var content: () -> Content

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 18) {
                HStack(spacing: 14) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 14)
                            .fill(Color.ink)
                            .offset(y: 4)
                        RoundedRectangle(cornerRadius: 14)
                            .fill(game.accentSoft)
                            .overlay(RoundedRectangle(cornerRadius: 14).strokeBorder(Color.ink, lineWidth: 3))
                        Image(systemName: game.symbol)
                            .font(.system(size: 24))
                            .foregroundStyle(Color.ink)
                    }
                    .frame(width: 54, height: 54)
                    VStack(alignment: .leading, spacing: 2) {
                        Text(game.title)
                            .font(.heading(26, weight: .semibold))
                            .foregroundStyle(Color.ink)
                        Text(game.instructions)
                            .font(.system(size: 13))
                            .foregroundStyle(Color.mutedInk)
                    }
                }

                picker()

                if engine.phase == .playing, !engine.words.isEmpty {
                    ScoreBar(
                        unit: unit,
                        index: engine.index,
                        total: engine.words.count,
                        score: engine.score,
                        streak: engine.streak,
                    )
                }

                VStack {
                    if engine.phase == .done {
                        RoundSummaryView(
                            score: engine.score,
                            total: engine.words.count,
                            bestStreak: engine.bestStreak,
                            missed: engine.missedWords,
                            summaryText: summaryText,
                            onRestart: onRestart,
                        )
                    } else {
                        content()
                    }
                }
                .padding(18)
                .frame(maxWidth: .infinity)
                .background(
                    RoundedRectangle(cornerRadius: 18)
                        .fill(Color.white)
                        .shadow(color: Color.ink.opacity(0.08), radius: 6, y: 3)
                )
                .overlay(alignment: .top) {
                    UnevenRoundedRectangle(
                        topLeadingRadius: 18,
                        topTrailingRadius: 18,
                    )
                    .fill(game.accent)
                    .frame(height: 7)
                }
            }
            .padding(16)
        }
        .background(Color.paper)
        .navigationBarTitleDisplayMode(.inline)
    }
}

// MARK: - Simple flow (wrap) layout

struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        arrange(proposal: proposal, subviews: subviews).size
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        let arrangement = arrange(proposal: proposal, subviews: subviews)
        for (subview, position) in zip(subviews, arrangement.positions) {
            subview.place(
                at: CGPoint(x: bounds.minX + position.x, y: bounds.minY + position.y),
                proposal: .unspecified,
            )
        }
    }

    private func arrange(proposal: ProposedViewSize, subviews: Subviews) -> (size: CGSize, positions: [CGPoint]) {
        let maxWidth = proposal.width ?? .infinity
        var positions: [CGPoint] = []
        var x: CGFloat = 0
        var y: CGFloat = 0
        var rowHeight: CGFloat = 0
        var totalWidth: CGFloat = 0

        for subview in subviews {
            let size = subview.sizeThatFits(.unspecified)
            if x > 0, x + size.width > maxWidth {
                x = 0
                y += rowHeight + spacing
                rowHeight = 0
            }
            positions.append(CGPoint(x: x, y: y))
            x += size.width + spacing
            rowHeight = max(rowHeight, size.height)
            totalWidth = max(totalWidth, x - spacing)
        }
        return (CGSize(width: totalWidth, height: y + rowHeight), positions)
    }
}
